tests/lp_hash_table_test.cpp

```diff
@@ -64,3 +65,36 @@ bool test_many_key_insertion() {
 
   return true;
 }
+
+/*
+ * NOTE: this test will fail if the compiler implementation interns strings.
+ * IE if two `const char *` have the same address in memory this test will fail.
+ */
```

Comment:
Fair point. I could do something like this
```c
  const char *k1 = "key1";
  char *k2 = (char *)malloc(5 * sizeof(char));
  std::memcpy(k2, k1, sizeof(k2));
```
but I think that this defeats the purpose of this test because it is a workaround for an implementation specific "problem". Shoud I remove this case all together or conditionally run it if the two pointers don't point at the same location?

Reply:
> but I think that this defeats the purpose of this test because it is a workaround for an implementation specific "problem".

Oh no, it doesn't! :) As you said, this is an implementation detail. The tests should not be based
on things that are implementation-dependent! We should try our best against things like that.

Reply:
> ```c
>   const char *k1 = "key1";
>   char *k2 = (char *)malloc(5 * sizeof(char));
>   std::memcpy(k2, k1, sizeof(k2));
> ```

In this, you may want to try `strdup()`. This is wrong btw because the `sizeof(k2)` in
`memcpy()` should be 5. `sizeof(k2)` is the size of the _pointer type_, not the memory it points to. That is always 4 or 8 depending on the target.
Last, it's better to do an `strlen` rather than `5`.

Reply:
The point was the dynamic allocation yes. So what I should do is test both implementation cases right?

Reply:
The point is to have different memory locations. So:
> So what I should do is test both implementation cases right?

No. That is because let's say you want to test X. You don't try to test for all the different
implementation cases. You're just trying to know about them, so that you try to not be dependent in them. For example, in this case, you want to test that different pointers with the same contents are saved in different places. So, you just try to find a way to _always_ have different memory with the same contents, no matter the implementation. That is achieved e.g. by allocating dynamically.

Reply:
Yeah it was clear from the start I just rushed it at this point I guess.

Reply:
Good :)